doc/man7/EVP_KDF_X942.pod

```diff
@@ -0,0 +1,156 @@
+=pod
+
+=head1 NAME
+
+EVP_KDF_X942 - The X9.42-2001 asn1 EVP_KDF implementation
+
+=head1 DESCRIPTION
+
+The EVP_KDF_X942 algorithm implements the key derivation function (X942KDF).
+X942KDF is used by Cryptographic Message Syntax (CMS) for DH KeyAgreement, to
+derive a key using input such as a shared secret key and other info. The other
+info is DER encoded data that contains a 32 bit counter.
+
+=head2 Numeric identity
+
+B<EVP_KDF_X942> is the numeric identity for this implementation; it
+can be used with the EVP_KDF_CTX_new_id() function.
+
+=head2 Supported controls
+
+The supported controls are:
+
+=over 4
+
+=item B<EVP_KDF_CTRL_SET_MD>
+
+This control works as described in L<EVP_KDF_CTX(3)/CONTROLS>.
+
+=item B<EVP_KDF_CTRL_SET_KEY>
+
+This control expects two arguments: C<unsigned char *secret>, C<size_t secretlen>
+
+The shared secret used for key derivation.  This control sets the secret.
+
+EVP_KDF_ctrl_str() takes two type strings for this control:
+
+=over 4
+
+=item "secret"
+
+The value string is used as is.
+
+=item "hexsecret"
+
+The value string is expected to be a hexadecimal number, which will be
+decoded before being passed on as the control value.
+
+=back
+
+=item B<EVP_KDF_CTRL_SET_OTHER_INFO>
+
+This control expects two arguments: C<unsigned char *info>, C<size_t infolen>
+
+A der encoded object containing fields such as algorithm id, counter,
+optional partyinfo and suppPubInfo. This control sets the other info.
```

Comment:
I find it kind of weird that you are expected to encode this DER object yourself rather than passing the values contained in it to the KDF, and have the KDF do the encoding of the DER for you.

Reply:
Yes its weird but I do it for a very good reason.. I dont want asn1 code inside when the kdf code gets moved into a provider.

Reply:
Hmmm....but its not a very nice interface. Could we encode the ASN.1 "by hand"? That was my understanding of our plan for the limited cases where we need ASN.1 inside the module boundary.

Reply:
I guess it could, or we could supply an external function to do it..  this is really a dh key agreement related algorithm - so its not general purpose. Maybe the embedded encode is the best way (it may require passing a few more control values in)

Reply:
As this algorithm will only exist in the default provider I have assumed that it can access public API's of libcrypto. So I have moved the ASN1 encoding code inside the EVP_KDF code. We should not need to use a custom ASN1 encoder for this case.